Document and add math functions

krader1961 opened the pull request:
In addition to documenting the `math:` module introduced by commit 2d8a045c
two months ago this also:

a) restructures the code to match that of the `re:` and `str:` modules, and

b) adds many more useful constants and functions.

I don't think this completely addresses issue #727 so I'm marking this
as related to, rather than fixing, that issue.

Related #727

krader1961 commented on pkg/edit/complete/complete_test.go:
```diff
@@ -110,7 +110,7 @@ func TestComplete(t *testing.T) {
 
 	allFileNameItems := []completion.Item{
-		fc("a.exe", " "), fc("d" + string(os.PathSeparator), ""), fc("non-exe", " "),
+		fc("a.exe", " "), fc("d"+string(os.PathSeparator), ""), fc("non-exe", " "),
```
This change is the result of `go fmt`. Since it was the only line in the existing code that was reformatted I decided to just include it even though it is not otherwise relevant to this PR.

krader1961 commented:
Ping? @xiaq, If you're not in the middle of commenting on this P.R. I'd like to rebase it on the master branch to resolve conflicts. Not to mention ensuring it builds on all of my VMs including my recently created Windows 10 plus Msys2 environment.

xiaq commented on pkg/eval/vals/equal.go:
```diff
@@ -22,7 +25,17 @@ func Equal(x, y interface{}) bool {
 		return x == y
 	case float64:
-		return x == y
+		if math.IsNaN(x) && math.IsNaN(y.(float64)) {
```
I don't like the idea of making floating point equality smart; best to remove this change from the PR.

krader1961 replied:
Fair enough. This was mostly a Go learning experience and a "what if...". But as the subsequent comment explains this was also an expedient way to test that NaN is output in unit tests. I'll revert this and give more thought to the unit test case.

xiaq commented on pkg/program/shell/runtime.go:
```diff
@@ -12,6 +12,7 @@ import (
 	"github.com/elves/elvish/pkg/daemon"
 	"github.com/elves/elvish/pkg/eval"
 	daemonmod "github.com/elves/elvish/pkg/eval/daemon"
+	elvish_math "github.com/elves/elvish/pkg/eval/math"
```
Do not use underscores in identifiers. Follow the pattern above and call the import alias `mathmod`.

krader1961 replied:
Okay

krader1961 replied:
Wouldn't it be more consistent to use snake-case to mimic the var naming convention; i.e., `mathMod`?

xiaq commented on website/ref/math.md:
```diff
@@ -0,0 +1,16 @@
+# Introduction
+
+The `math:` module provides mathematical functions and
+constants. Everyplace a `$float64` is referenced below that data type
```
This explanation is no longer needed. Just add a reference to It now suffices to just refer to https://elv.sh/ref/builtin.html#commands-that-operate-on-numbers at the end of the next paragraph (row 14).

krader1961 replied:
Done

krader1961 commented:
I addressed the above issues, rebased on master, and replaced `$float64` mentions in the command documentation with `$number`. Note that I've replaced the problematic code in `vals.Equal()` with equivalent code in `testutils.matchOut()`.

xiaq commented on pkg/eval/vals/equal.go:
```diff
@@ -22,7 +25,11 @@ func Equal(x, y interface{}) bool {
 		return x == y
 	case float64:
-		return x == y
+		if math.Signbit(x) == math.Signbit(y.(float64)) &&
```
Let's just stick to Go's standard `==` semantics here.

krader1961 replied:
We can't because your code does not stick to Go's standard `==` semantics. The code I added is compatible with Go's standard `==` semantics for infinities while still supporting the +/- epsilon of your code. Without this change if I replace your

```
return math.Abs(x-y.(float64)) <= float64EqualityThreshold
```

with

```
return x == y.(float64)
```

then my unit test passes but this failure occurs:

```
ok      github.com/elves/elvish/pkg/eval/daemon 0.300s
--- FAIL: TestMath (0.09s)
    --- FAIL: TestMath/math:sin_3.141592653589793 (0.00s)
        testutils.go:143: got value out [1.2246467991473515e-16], want [0]
    --- FAIL: TestMath/math:tan_1 (0.00s)
        testutils.go:143: got value out [1.557407724654902], want [1.5574077246549023]
    --- FAIL: TestMath/math:tan_3.141592653589793 (0.00s)
        testutils.go:143: got value out [-1.2246467991473515e-16], want [0]
```

krader1961 replied:
/facepalm :smile: I'm the one who added the epsilon test. Okay, I'll revert that portion of the change.

krader1961 commented on pkg/eval/math/math.go:
````diff
@@ -0,0 +1,288 @@
+// ```
+
+// Ns is the namespace for the math: module.
+var Ns = eval.NewNs().AddGoFns("math:", fns)
````
I like, and use, method chaining in a lot of Python code I deal with. I am not a fan of lengthy method chaining. It tends to obscure what is happening. Which makes it harder to understand the code. I appreciate that at the moment the alternative formulation you propose is easy to understand. The problem is what happens when additional constants, methods, etc., are chained together. Which will almost certainly occur. You've objected to the use of `init()` functions before but I don't understand why since, AFAICT, they don't introduce any performance or correctness problems when used in this manner. The mixing of `eval.Ns{...}` declarations with `eval.NewNs()`

Nonetheless, to facilitate getting this merged I've made the suggested change.

P.S., There are just four uses of `NewNs()`. Compared to twenty-four uses of `Ns{}`. Should the former be converted to the latter form?

xiaq replied:
**Re use of method chaining**:

The problem with `init` is that it does not detect cyclic dependencies in initialization. Compare the following two pieces of code:

Using `init`:

```go
// In file a.go
var a int
func init() { a = b + 1 }

// In file b.go
var b int
func init() { b = a + 1}
```

Not using `init`:

```go
// In file a.go
var a int = b + 1

// In file b.go
var b int = a + 1 
```

Go will correctly point out that the second version has an initialization loop, but it will happily compile the first version. Now, what should the values of `a` and `b` be? :)

It is often a matter of trade-off, of course. Sometimes using `init` is much cleaner than the alternative that avoids it. But in this case, my call is that method chaining is the lesser evil here.

**Re use of NewNs**: Yes, they can be all replaced by `Ns{}`.

krader1961 replied:
I understand your point about an indeterminate initialization sequence. That is inherent in something like Go's module level `init()` mechanism. I was not aware that the compiler would error out if that happens in an example such as you showed. I'll explore that feature. Nonetheless, I still maintain that using the `init()` mechanism is clearer in this situation for non-trivial initialization and, when used in a stylized fashion with an understanding of the possible indeterminacy, is easier to understand.

krader1961 commented:
Sorry about the squashed force push. But I wanted to resolve the conflict with the recent changes to the master branch. Despite that this update only addresses the two most recent issues:

1) Restoring the behavior of `== (float64 $x) (float64 $y)` to use only Go's built-in equality without applying an epsilon. That logic has been moved into `matchOut()` in the *testutils.go* module.

2) Removing the use of `init()` in *math.go*.

xiaq commented on pkg/eval/math/math_test.go:
```diff
@@ -0,0 +1,104 @@
+		That(`math:trunc (float64 Inf)`).Puts(math.Inf(1)),
+		That(`math:trunc (float64 NaN)`).Puts(math.NaN()),
+
+		// That(`math:log 2.71828182845904523536028747135266249775724709369995`).Puts(1.0),
```
Do not commit commented code; just remove it.

krader1961 replied:
Yes, that was an oversight from a prior experiment for improving the test. And the result of that experiment was also flawed since the test should be using the elvish `$math:e` constant, rather than interpolating the Go `math:E` constant as you pointed out in a different comment. :smile:

xiaq commented on pkg/eval/math/math_test.go:
```diff
@@ -0,0 +1,104 @@
+		That(`math:trunc (float64 NaN)`).Puts(math.NaN()),
+
+		// That(`math:log 2.71828182845904523536028747135266249775724709369995`).Puts(1.0),
+		That(fmt.Sprintf(`math:log %g`, math.E)).Puts(1.0),
```
Use `math:log $math:e`.

Same for the test cases that use `math.Pi`.

krader1961 replied:
Done

xiaq commented on pkg/eval/vals/equal.go:
```diff
@@ -4,6 +4,8 @@ import (
 	"reflect"
 )
 
+const float64EqualityThreshold = 1e-9
```
Remove

krader1961 replied:
Even if I implement `Approximate()`, as you suggested in a different comment, we still need this constant or something like it. So I don't understand this comment.

xiaq replied:
You already the same constant in testutils.go, with a different value :)

krader1961 replied:
Gah! Sorry, I thought I had removed that already.

xiaq commented on website/ref/math.md:
```diff
@@ -0,0 +1,11 @@
+The `math:` module provides mathematical functions and constants.
+
+Function usages are given in the same format as in the reference doc for the
+[builtin module](builtin.html). In particular, the discussion of [Commands
```
> In particular, ...

* Write complete sentences to make the meaning unambiguous.

* Only use title case when explicitly referring to a section (such as "see section Number Type"). In this case you are referring to the *content* of a section.

In this case, write:

`In particular, all the commands in this module conform to the pattern of [commands that operate on numbers](...).`

xiaq commented on pkg/eval/testutils.go:
```diff
@@ -223,6 +226,29 @@ func matchOut(want, got []interface{}) bool {
 		return false
 	}
 	for i := range got {
+		// Equality of some data types needs to be special-cased in unit tests.
```
I like how you have moved the approximate comparison into `matchOut`, but I'd prefer to make it even more explicit by doing the following:

1. Add a new type, `type Approximately struct{ f float64 }`.

2. When `want[i]` has that type, compare with `got[i]` with an epsilon.

3. Change the unit tests that need to compare float64 approximately from:

    ```go
    That("some code").Puts(1.0)
    ```

    to

    ```go
    That("some code").Puts(Approximately{1.0})
    ```

    Most test cases will *not* need approximate comparison. You can simply run the tests, see which ones are failing, and change those to use `Approximately`.

This way, test cases that need precise will continue to do so, and test cases that need approximate comparison are very explicit.

krader1961 replied:
> Most test cases will not need approximate comparison. You can simply run the tests, see which ones are failing, and change those to use Approximately.

Yes, most test cases do not need an approximate comparison. But that is simply because the existing tests have predictable bit patterns. For example, the constant `1` has a predicable float64 bit pattern. If every time a test involving float64 values fails due to a small epsilon, and the code is changed to use `Approximately()`, what is the point?

krader1961 replied:
I'm going to go ahead and implement your suggestion but note that a test such as

```
That(`math:cos 1`).Puts(fmt.Sprintf("%g", math.Cos(1.0))),
```

still fails because the default invocation of `vals.Equal()` doesn't handle the inconsequential difference between the two values.

xiaq replied:
> But that is simply because the existing tests have predictable bit patterns. 

It's because Elvish needs to guarantee that certain floating point number operations *have* accurate results. For example, 1.0 + 2.0 needs to be exactly 3.0 (unlike 0.1 + 0.2, which is approximately 3.0). Since Elvish does not offer an integer type yet this has to be part of what Elvish promises.

> If every time a test involving float64 values fails due to a small epsilon, and the code is changed to use Approximately(), what is the point?

I suggested that you add `Approximately` to failing tests as a quick way to fix the tests, but that's *not* the criteria for whether a test should have precise or approximate in the first place. See comment above.

> ... still fails because the default invocation of vals.Equal() doesn't handle the inconsequential difference between the two values.

The difference *is* consequential because the two values *are* different. They have different JSON presentations for example, and the tests should be precise in which form it expects.

In general, I feel you are missing an important distinction between what *an average user would expect* and *what Elvish should promise*. It's completely OK for an Elvish user to think "oh floating point numbers are messy, better to not expect anything accurate", but the same attitude is counter-productive for an Elvish developer. You need to think about all the edges cases where the subtle difference can translate into a bigger one and be as rigorous as possible.

xiaq commented on pkg/eval/math/math.go:
````diff
@@ -0,0 +1,286 @@
+// $math:e
+// ```
+//
+// The value of E: 2.718281.... This variable is read-only.
````
Write *e* with a lower case and italic to be consistent with how it is usually written.

Linking to Wikipedia would also be nice.

krader1961 replied:
Done

xiaq commented on pkg/eval/math/math.go:
````diff
@@ -0,0 +1,286 @@
+// $math:pi
+// ```
+//
+// The value of Pi: 3.141592.... This variable is read-only.
````
I'd write it as *π*, using the Greek letter in italic, and link to Wikipedia.

krader1961 replied:
Done

xiaq commented on pkg/eval/math/math.go:
````diff
@@ -0,0 +1,286 @@
+// math:abs $number
+// ```
+//
+// Compute the absolute value of a number. Example:
````
Use third-person singular form of the verb, "computes".

In case you haven't noticed, there is now a [style guide](https://github.com/elves/elvish/blob/master/CONTRIBUTING.md#documenting-builtin-functions-and-variables) for the "elvdoc" comments.

Same for functions below.

krader1961 replied:
Okay, but how to square the circle? That is, what about the commands like `math:round` which say `Output the nearest integer` rather than `Outputs...`?

krader1961 replied:
Ah, never mind, I see you made another comment about s/Output/Outputs/.

xiaq commented on pkg/eval/math/math.go:
````diff
@@ -0,0 +1,286 @@
+// math:ceil $number
+// ```
+//
+// Compute the floor value of a number. Example:
````
This should say "ceiling".

Also, I like how https://godoc.org/math#Ceil actually explains the definition of ceiling and lists the special cases.

krader1961 replied:
Yes, stupid copy/paste mistake. I will add a "Read the Go documentation..." link but that begs the question of when such links should be present.

xiaq commented on pkg/eval/math/math.go:
````diff
@@ -0,0 +1,286 @@
+// math:floor $number
+// ```
+//
+// Compute the floor value of a number. Example:
````
See comment on `ceil` above.

krader1961 replied:
Done

xiaq commented on pkg/eval/math/math.go:
````diff
@@ -0,0 +1,286 @@
+//
+// ```elvish
+// math:is-inf $number $sign
+// ```
````
This function could have a more [Elven](https://elv.sh/ref/name.html#related-words) API by taking the `$sign` argument as an option instead, and defaulting to 0. This way it can be used as follows:

```
math:is-inf $n # tests if $n is +inf or -inf
math:is-inf $n &sign=-1 # tests if $n is -inf
math:is-inf $n &sign=1 # tests if $n is +inf
```

I also like how https://godoc.org/math#IsInf explains the use of `$sign` in separate sentences - it's more readable that way.

krader1961 replied:
SGTM

xiaq commented on pkg/eval/math/math.go:
````diff
@@ -0,0 +1,286 @@
+// math:is-nan $number $sign
+// ```
+//
+// If the number is not-a-number (NaN) output `$true` else `$false`.
````
Say

> Tests whether the number is NaN.

xiaq commented on pkg/eval/math/math.go:
````diff
@@ -0,0 +1,286 @@
+// math:trunc $number
+// ```
+//
+// Output the integer value of the float64.
````
Say

> Outputs the integer value of $number.

xiaq commented:
Please resolve the merge conflict and rebase to get accurate CI results.

krader1961 commented:
Rebase done, all tests are passing, and I don't see any obvious documentation problems after running `cd website; make gen` and viewing the results in my web browser.

xiaq commented:
Going to merge this. @krader1961, thank you for your patience!
